File: app/src/main/java/com/example/convertingtokotlin/Repository.java
package com.example.convertingtokotlin;

import java.util.ArrayList;
import java.util.List;

public class Repository {

    // list of users
    private List<User> users = null;

    // singleton
    private static Repository INSTANCE = null;
    public static Repository getInstance() {

        if (INSTANCE == null) {
            synchronized (Repository.class) {
                if (INSTANCE == null) INSTANCE = new Repository();
            }
        }
        return INSTANCE;
    }

    // keeping the constructor private to enforces the usage of getInstance
    private Repository() {

        // create some users
        User user1 = new User("Jane", "");
        User user2 = new User("John", null);
        User user3 = new User("Anne", "Doe");

        // create list of users
        users = new ArrayList();
        users.add(user1);
        users.add(user2);
        users.add(user3);
    }

    // get users
    public List<User> getUsers() {
        return users;
    }

    // get formatted user names
    public List<String> getFormattedUserNames() {

        // create list of user names
        List<String> userNames = new ArrayList<>(users.size());

        // build list of names
        for (User user : users) {

            String name;

            if (user.getLastName() != null) {

                if (user.getFirstName() != null) name = user.getFirstName() + " " + user.getLastName();
                else name = user.getLastName();
            }

            else if (user.getFirstName() != null) name = user.getFirstName();
            else name = "Unknown";

            userNames.add(name);
        }

        return userNames;
    }
}

File: app/src/main/java/com/example/convertingtokotlin/Repository.kt
package com.example.convertingtokotlin

// extension property
val User.formattedName: String
    get() {
        return if (lastName != null) {
            if (firstName != null) "$firstName $lastName"
            else lastName ?: "Unknown"
        } else firstName ?: "Unknown"
    }

// object is singleton
object Repository {

    // get list of users
    private val _users = mutableListOf<User>()
    val users: List<User>
        get() = _users

    // get formatted user names
    val formattedUserNames: List<String>
        get() {
            return _users.map{ user -> user.formattedName }
        }

    // initialize some example users
    init {

        // create some users
        val user1 = User("Jane", "")
        val user2 = User("John", null)
        val user3 = User("Anne", "Doe")

        // create list of users using scope function
        _users.apply {
            add(user1)
            add(user2)
            add(user3)
        }
    }
}